src/store/reducers/query/useQueriesHistory.ts

```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+
+import {
+    useEventHandler,
+    useSetting,
+    useTypedDispatch,
+    useTypedSelector,
+} from '../../../utils/hooks';
+import {SETTING_KEYS} from '../settings/constants';
+
+import {changeUserInput, selectQueriesHistoryFilter} from './query';
+import type {QueryInHistory, QueryStats} from './types';
+import {getQueryInHistory} from './utils';
+
+const MAXIMUM_QUERIES_IN_HISTORY = 20;
+
+export function useQueriesHistory() {
+    const dispatch = useTypedDispatch();
+    const queriesFilter = useTypedSelector(selectQueriesHistoryFilter);
+
+    const [savedHistoryQueries, saveHistoryQueries] = useSetting<QueryInHistory[]>(
+        SETTING_KEYS.QUERIES_HISTORY,
+    );
+
+    const [historyQueries, setQueries] = React.useState<QueryInHistory[]>([]);
+    const [historyCurrentIndex, setCurrentIndex] = React.useState(-1);
+
+    React.useEffect(() => {
```

Comment:
is it possible to check this before saving (not in useEffect)?

Reply:
We can receive `undefined` value from meta, if setting was never set before, so this condition is needed to migrate safely after.

As for setting or updating value, it will be always valid and defined, since I create new value based on current, and current value should be always valid because of this `useEffect`

Reply:
Agreed in PM to get rid of this effect. Also agreed recalculate actual values and slice over limit queries in handlers